databroker/broker.py

```diff
@@ -880,3 +902,178 @@ def _munge_time(t, timezone):
+                                         provenance=provenance)
+                    yield 'start', new_start_doc
+
+                elif name == 'descriptor' and doc.get('name') == stream_name:
+                    if run_start_uid is None:
+                        raise RuntimeError("Received EventDescriptor before "
+                                           "RunStart.")
+                    descriptor_uid = doc_or_uid_to_uid(doc)
+                    new_data_keys = dict(doc['data_keys'])
+                    for k, v in new_data_keys.items():
+                        new_data_keys[k].update(v)
+                    new_descriptor_uid = str(uuid.uuid4())
+                    new_descriptor = dict(uid=new_descriptor_uid,
+                                          time=time.time(),
+                                          run_start=run_start_uid,
+                                          data_keys=new_data_keys,
+                                          name=stream_name)
+                    yield 'descriptor', new_descriptor
+
+                elif (name == 'event' and
+                      doc_or_uid_to_uid(doc['descriptor']) == descriptor_uid):
+                    if run_start_uid is None:
+                        raise RuntimeError("Received Event before RunStart.")
+                    try:
+                        new_event = dict(doc)
+                        # We need a selectively deeper copy since we will
+                        # mutate the contents of new_event['data'].
+                        new_event['data'] = dict(new_event['data'])
+                        new_event['uid'] = str(uuid.uuid4())
+                        new_event['descriptor'] = new_descriptor_uid
+                        for data_key in new_data_keys:
+                            value = doc['data'][data_key]
+                            new_event['data'][data_key] = f(value, **kwargs)
+                        yield 'event', new_event
+                    except Exception as e:
+                        new_stop = dict(uid=str(uuid.uuid4()),
+                                        time=time.time(),
+                                        run_start=run_start_uid,
+                                        exit_status='failure')
+                        yield 'stop', new_stop
+                        raise
+
+                elif name == 'stop':
+                    if run_start_uid is None:
+                        raise RuntimeError("Received RunStop before RunStart.")
+                    new_stop = dict(uid=str(uuid.uuid4()),
+                                    time=time.time(),
+                                    run_start=run_start_uid,
+                                    exit_status='success')
+                    descriptor_uid = None
+                    run_start_uid = None
+                    yield 'stop', new_stop
+        return inner
+    return outer
+
+
+def header_io(db_in, db_out):
```

Comment:
I'd be tempted to remove this. As a user, I am fine with being forced to think in terms of streams for more complex tasks. I can convert whatever object it is I use to store data to a compatible stream no problem. I think if you add this, you might force too many standards on yourselves.
I think @danielballan and @tacaswell  you were mentioning this today (correct me if I'm wrong), that you want to make sure we understand documents and see headers as a convenient cul-de-sac (sorry I like alliteration).
What do you think? Might require longer discussion.